### PR TITLE
Move implementations - learning layers

### PR DESCRIPTION
**Pull request by benson31**
In an ideal world, we'd be able to move all the implementations into non-headers (since they aren't _real_ templates and we only allow the ETI'd parameter combinations). But some of the classes have the implementation split between cpp and cu files, both needing the implementations. For those files, I've just moved the code out of the main class template definition, but it's still in the main header. A more robust refactoring would move the code to a `<layer_name>_impl.hpp` file, probably in the `src/` tree. I've elected to not do this today; let me know if you'd prefer that approach. Relevant layers are, e.g., embedding, channelwise/entrywise scale bias.

This doesn't modify any functionality. It just moves things around and defines some typedefs.

### CODE REVIEW

File: include/lbann/layers/learning/entrywise_scale_bias.hpp [hunk 1]
```diff
@@ -70,124 +70,161 @@ class entrywise_scale_bias_layer : public data_type_layer<TensorDataType> {
...
-          << "has misaligned input and weights matrices";
-      LBANN_ERROR(err.str());
-    }
+  /** @brief Objective function gradient w.r.t. weights. */
+  std::unique_ptr<AbsDistMatrixType> m_weights_gradient;
 
-#endif // 0
+};
 
+// Implementation
```
Comment (bvanessen) [inline include/lbann/layers/learning/entrywise_scale_bias.hpp:104]: As with channelwise scale bias, why is this not moved to an impl file or cpp.

File: include/lbann/layers/learning/channelwise_scale_bias.hpp [hunk 2]
```diff
@@ -75,84 +75,110 @@ class channelwise_scale_bias_layer : public data_type_layer<TensorDataType> {
...
 
 private:
 
-  /** Objective function gradient w.r.t. weights. */
+  /** @brief Objective function gradient w.r.t. weights. */
   std::unique_ptr<AbsDistMatrixType> m_weights_gradient;
 
 };
 
+// Implementation
```
[the remark this answers is not in the document]
Comment (timmoon10) [reply include/lbann/layers/learning/channelwise_scale_bias.hpp:106]: There are two compilation units that require access to this implementation (`channelwise_scale_bias.cpp` and `channelwise_scale_bias.cu`). From what I gather, the right thing to do would be to put this into a header `channelwise_scale_bias_impl.hpp` that lives in the source directory and is included by the source files. I think keeping it here for now is a nice compromise though between good practice, keeping the directory structure simple, and laziness.

[the remark this answers is not in the document]
Comment (benson31) [reply include/lbann/layers/learning/channelwise_scale_bias.hpp:106]: What @timmoon10 said.

[the remark this answers is not in the document]
Comment (bvanessen) [reply include/lbann/layers/learning/channelwise_scale_bias.hpp:106]: I see.  Thanks.

**Pull request merged**